Feature: npm module

hyounesy opened the pull request:
making all changes as a pull request to facilitate code review and commenting

hyounesy committed:
added template-npm-package-master

hyounesy committed:
added RcSepp/GlobalView.js

hyounesy committed:
added global view modules and browserify+beefy.
copied example1.html to index.html and it is working

hyounesy committed:
temporary: moved external modules to src_lib and copy them directly to lib using cpx to prevent them from going through cross-env.
todo: remove them as require() is added

hyounesy committed:
temporarily include /lib in git to be able to diff the changes. todo: remove before push

hyounesy committed:
moved out webgl-utils from src_libs to src

hyounesy committed:
moving example1

hyounesy committed:
added example2 and example3

hyounesy committed:
refactoring examples 4-7 + index

hyounesy committed:
added examples 4, 5

hyounesy committed:
added example6

hyounesy committed:
added example7 and example8 (previously index.html, main.js)
todo: benchmark doesn't work (needs global variable globalView)

hyounesy committed:
work in progress: packaging all globalView modules.

hyounesy committed:
moving all external dependencies to use npm. fixed example5.

hyounesy committed:
refactored all examples to use the bundled global-view

hyounesy committed:
ported examples to use webpack (keeping browserify for the time being)

hyounesy committed:
getting it ready for a trial npm push. updated README and package.

hyounesy committed:
enabling eslint. adding aslant-nibble,
fixed: space-before-function-paren

hyounesy committed:
eslint: fixed space-in-parens

hyounesy committed:
eslint: no-tabs

hyounesy committed:
eslint: no-tabs

hyounesy committed:
eslint: brace-style

hyounesy committed:
eslint: no-trailing-spaces

hyounesy committed:
eslint: quotes

hyounesy committed:
removed colormap images and embedded them as base64 in colormap.js

hyounesy committed:
minor cleaning up.
example8: renamed vars to make it consistent with the other examples

hyounesy committed:
no more warnings when pointColor is a colormap name

hyounesy committed:
useless intermediate files...

hyounesy committed:
added example9. it is the work in progress code to add unit tests.

hyounesy committed:
making the tests modular

hyounesy committed:
adding tests

hyounesy committed:
removing intermediate files

hyounesy committed:
updated gitignore

hyounesy committed:
added more tests

hyounesy committed:
fixed the tests for cluster and density options.

hyounesy committed:
added tests for histogram options.
After adding 17th test, get the message: global-view.js:5969 WARNING: Too many active WebGL contexts. Oldest context will be lost.
So, have to limit the number of interactive tests (e.g. click to make a plot interactive).

hyounesy committed:
added test for thumbnails

GabeMedrash commented on .gitignore:
```diff
@@ -0,0 +1,10 @@
+.vscode/
+lib/test
+examples/*.min.js
+src/test/*.min.js
```
All build artifacts should be ignored, including `lib/` and `dist/`.

hyounesy replied:
Done

GabeMedrash commented on GlobalView.js-master/LICENSE:
```diff
@@ -0,0 +1,21 @@
+MIT License
```
I'm guessing this was a provisional measure as you were working, but, just to say it, the whole `GlobalView.js-master` directory should eventually be removed.

hyounesy replied:
Done

GabeMedrash commented on globalView.css:
```diff
@@ -0,0 +1,71 @@
+* {
```
What is this used for? If it's for examples, it should be moved under the `examples` directory.

hyounesy replied:
Done. Moved under examples.

GabeMedrash commented on index.html:
```diff
@@ -0,0 +1,22 @@
+<!DOCTYPE HTML>
```
This should also be moved under the `examples` directory.

hyounesy replied:
Done.

GabeMedrash commented on package.json:
```diff
@@ -0,0 +1,89 @@
+  "module": "es/index.js",
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/RcSepp/GlobalView.js.git"
```
This should be updated, although we'll have to talk to Megan/Graham about the eventual ownership of this repo (i.e., whether it will end up under some Allen Institute for Cell Science GitHub org).

hyounesy replied:
Yes. Please update as you see fit.

hyounesy committed:
moving example?.min.js to bin

GabeMedrash commented on package.json:
```diff
@@ -0,0 +1,89 @@
+    "transpile:commonjs": "cross-env BABEL_ENV=commonjs babel src --out-dir lib",
+    "transpile:es": "cross-env BABEL_ENV=es babel src --out-dir es",
+    "test": "",
+    "statstest": "cross-env BABEL_ENV=test mocha --require babel-polyfill --require babel-register src/**/*.test.js",
```
The whole stats thing in the template was really just for example purposes. But I would like eventually for unit tests to be written that can be run with this exact script (`cross-env BABEL_ENV=test mocha --require babel-polyfill --require babel-register src/**/*.test.js`). The tests you've included are great and important as you make changes, but are significantly larger than simple unit tests--they seem to me to be functional tests for the entire package.

hyounesy replied:
For now, we decided to use snapshot testing in place of conventional unit testing as this project is more of an exceptional case. Most of the functions require  webgl context and canvas elements, and use WebGL Shaders which make them very difficult (if not currently impossible) to properly work on CLI. In addition are also async functions currently managed with Parallel library (spawn).
As the main purpose for the tests is to help with future development, we believe the snapshot tests are better suited for the purpose.

GabeMedrash commented on package.json:
```diff
@@ -0,0 +1,89 @@
+    "version": "npm run build && git add .",
+    "examples": "webpack --config ./examples/webpack.examples.config.js && webpack-dev-server --config ./examples/webpack.examples.config.js",
+    "tests": "webpack --config ./src/test/webpack.test.config.js && webpack-dev-server --config ./src/test/webpack.test.config.js",
+    "TODO_minify": "(TODO): cat bundle.js | uglifyjs -o bundle.js",
```
Not sure if this is being used or not, but minification should be accomplished with Webpack--no need for extra tooling. That can be accomplished by setting the `-p` flag in the call to webpack in the "bundle" task (https://webpack.js.org/api/cli/#shortcuts).

hyounesy replied:
Done. thanks for the help.

GabeMedrash commented on package.json:
```diff
@@ -0,0 +1,89 @@
+    "examples": "webpack --config ./examples/webpack.examples.config.js && webpack-dev-server --config ./examples/webpack.examples.config.js",
+    "tests": "webpack --config ./src/test/webpack.test.config.js && webpack-dev-server --config ./src/test/webpack.test.config.js",
+    "TODO_minify": "(TODO): cat bundle.js | uglifyjs -o bundle.js",
+    "nibble": "eslint-nibble src/ examples/"
```
Cool. Didn't know about eslint-nibble.

hyounesy replied:
Removed as the code is now fully linted and the standard eslint is probably enough for future development.

GabeMedrash commented on package.json:
```diff
@@ -0,0 +1,89 @@
+    "rimraf": "~2.6",
+    "save-pixels": "^2.3.4",
+    "seedrandom": "^2.4.3",
+    "tape": "^4.8.0",
```
Are you using `tape`? The template pulls in `mocha` and `chai` for your unit testing harness. If you are successfully using tape and prefer it, that's fine with me, but be sure to remove `mocha` and `chai`.

hyounesy replied:
removed

GabeMedrash commented on package.json:
```diff
@@ -0,0 +1,89 @@
+    "install": "^0.10.2",
+    "mocha": "~4.0",
+    "ndarray": "^1.0.18",
+    "npm": "^5.6.0",
```
You should not need to pull in `npm` as a development dependency. It is conventionally treated as a global (for example, in order to `npm install` in the first place you need to have npm on your machine). Sometimes npm/Node.js projects will declare an "engines" configuration in the package.json (https://docs.npmjs.com/files/package.json#engines), but this repo should not need one.

hyounesy replied:
removed.

GabeMedrash commented on package.json:
```diff
@@ -0,0 +1,89 @@
+    "webpack-dev-server": "^2.11.1"
+  },
+  "dependencies": {
+    "domready": "^1.0.8",
```
Any package used for examples and/or tests should be a "devDependency"

hyounesy replied:
done.

GabeMedrash commented on src/algorithm.js:
```diff
@@ -0,0 +1,1689 @@
+    // Copy constructor
+    //this.targetWidth = source.targetWidth;
+    //this.targetHeight = source.targetHeight;
+    this['maxExpectedRuntime'] = source['maxExpectedRuntime'];
```
Not sure if this was already the style, but in JS, using bracket syntax like this is unnecessary--unless there are non-alpha-numeric characters in the property name, use dot notation:
`this.maxExpectedRuntime = source.maxExpectedRuntime`;

Among other reasons, it can help with code hinting.

hyounesy replied:
all fixed.

GabeMedrash commented on src/algorithm.js:
```diff
@@ -0,0 +1,1689 @@
+export function ClusterMapOptions(source) {
+  if (source) {
+    this.densityMap = source.densityMap;
+    this['threshold'] = this.threshold = source.threshold;
```
Again, not sure if this was already the style in the codebase, but these extra assignments (I've seen them throughout `src`) are entirely unnecessary and should be removed.

hyounesy replied:
all fixed.

GabeMedrash commented on src/algorithm.js:
```diff
@@ -0,0 +1,1689 @@
+ * @param {ClusterMapOptions=} source If not null, creates a copy of source
+ */
+export function ClusterMapOptions(source) {
+  if (source) {
```
This can tell us if `source` is truthy, but doesn't tell us if it actually has the properties we're expecting it to have. For example, if `source === {}`, it would pass this guard statement, and would result in `this.densityMap` and `this.threshold` being set to `undefined`. It would be better to use defaults. For example:
Option 1:
```js
function Foo(obj) {
  const {
    bar = 3,
  } = obj;

  this.bar = bar;
}
```

Option 2:
```js
// in a utility file
export const isUndefined = (value) => value === undefined;
export const defaultTo = (value, default) => isUndefined(value) ? default : value;

// ...elsewhere
import { defaultTo } from "./path/to/utils";

// note below that I use a default parameter in case this constructor is called without a parameter
function Foo(obj = {}) {
  this.bar = defaultTo(obj.bar, 3);
}
```

hyounesy replied:
done (option2)

GabeMedrash commented on src/algorithm.js:
```diff
@@ -0,0 +1,1689 @@
+ * @return {Object} A 1D histogram
+ */
+export function computeHistogram(dataset, d, width) {
+  var n = dataset.length;
```
I'm 100% positive this was in the codebase from before, but two notes:
1. We shouldn't be using `var`; best practice to use `const` and `let`. 
2. We should never use single character variable names: they should describe what they hold/point to.

hyounesy replied:
ver -> let: done
single character variable names removed from all function parameters.

GabeMedrash commented on src/algorithm.js:
```diff
@@ -0,0 +1,1689 @@
+    maximum: maximum,
+    width: width, height: height,
+    transform: transform,
+    /*transformX: x => transform[0] * x + transform[1],
```
I think you already have this on your TODO list, but we should delete all commented out code.

hyounesy replied:
done.

GabeMedrash commented on src/algorithm.js:
```diff
@@ -0,0 +1,1689 @@
+  // Compute a measure of expected runtime
+  var expectedRuntime = 0;
+  var newBounds_l = Number.MAX_VALUE, newBounds_r = Number.MIN_VALUE, newBounds_t = Number.MAX_VALUE, newBounds_b = Number.MIN_VALUE;
+  tik();
```
It's really difficult for me to track what in this function is being used for actually computing a density map, and what is being used for calculating performance metrics. If any of it is being used for performance metrics, I consider that cruft and think it should be removed.

GabeMedrash commented on src/benchmarkDialog.js:
```diff
@@ -0,0 +1,221 @@
+//const libUtility = require('utility')
```
I think this file can be deleted? It was just for displaying the performance metrics, right?

hyounesy replied:
Used by the example8. Moved to examples.

GabeMedrash commented on src/colormap.js:
```diff
@@ -0,0 +1,344 @@
+
+// http://base64online.org/encode
+// http://textmechanic.co/Line-Length-Breaker.html
+const imageExhue =
```
100% agree with you that this should be a colorbrewer palette.

hyounesy replied:
will add this to the readme as a future development.

GabeMedrash commented on src/colormap.js:
```diff
@@ -0,0 +1,344 @@
+const libUtility = require('./utility.js');
+const libGraphics = require('./graphics.js');
+const libShaders = require('./shaders.js');
+const libGlMatrix = require('gl-matrix');
```
Some general comments about importing dependencies:
1. We should be using es modules (`import` and `export`, not `require`; see http://2ality.com/2014/09/es6-modules-final.html)
2. I don't see a strong reason to prefix everything with `lib`: there's no real danger of naming collision.
3. Conventional ordering in a block of import statements is as follows:
```js
// first, import third party (i.e., npm installed) packages
import glMatrix from 'gl-matrix';

// after a line separator, import modules from your own codebase, grouped by how far away the module is 
// from the current module
import grandParent from '../../gradparent';

import parent from '../parent';
import * as parent2 from '../parent2';

import { sibling } from './sibling';
```

hyounesy replied:
all done.

GabeMedrash commented on src/dataset.js:
```diff
@@ -0,0 +1,841 @@
+ * @param {number} nc Number of columns (dimensions) of the dataset
+ * @param {function(Dataset)} onload Event handler, called after the dataset was created
+ */
+export function RandomDataset(n, nc, onload) {
```
Is this used for anything other than examples or benchmarking? If not, should be deleted.

hyounesy replied:
I think it is a useful feature to have specially during debugging and development, plus it is a pretty tiny code. It is now refactored to a separate class, so I will let you decide whether you would like to keep or remove it.

GabeMedrash commented on src/formulaCompiler.js:
```diff
@@ -0,0 +1,638 @@
+const libUtility = require('./utility.js')
+
+export var FormulaCompiler = {
```
What is this? Is there any possible way we can delete it? Looking over this file makes me think this library is trying to do WAY too much.

hyounesy replied:
This is to allow custom formula based properties to be compiled and ran on the Shader. I think it is a pretty powerful feature to have, and I recommend keeping it until we are sure we don't want it anymore. For now, I have cleaned it up together with the rest of the code.

GabeMedrash commented on src/globalView.js:
```diff
@@ -0,0 +1,2179 @@
+
+var ENABLE_CONTINUOUS_RENDERING = false;
+var SHOW_FPS = false;
+var SIMULATE_LOW_FPS = false;
```
I'm guessing this stuff was for the benchmarking and examples? Can delete if so.

hyounesy replied:
These are to show the frame rate in continuous rendering. I am not sure if we want to remove them as it won't be trivial to compute fps from outside example codes.

GabeMedrash commented on src/webgl-utils.js:
```diff
@@ -0,0 +1,181 @@
+/**
```
Is this something that we can pull off of NPM instead? Or find something similar that is in NPM?

hyounesy replied:
Need to keep this as the npm version didn't quite work.

GabeMedrash commented on webpack.config.js:
```diff
@@ -0,0 +1,31 @@
+      },
+    ],
+  },
+  node: {
```
This makes me uncomfortable. Why do we need it?

hyounesy replied:
this is needed to get the Parallel library working with webpack: 
https://github.com/webpack/webpack/issues/744

GabeMedrash reviewed (commented):
Hamid, excellent work so far. You've taken an amorphous directory of code and given some good initial structure to it. I've left a number of inline comments on the PR. Some are general, some are specific. If you'd like, we can schedule a call to talk through them and if they fit into the scope of your contract.

One big point I want mention: I'd consider the tests you've written to be package-wide functional tests, which I totally see the use-case for (as a side-note, these are all failing for me). But I think the more significant thing is that I think time would be better spent writing unit tests to every extent possible. That way the code can evolve beyond the original spec.

Lastly, I tried to setup an example project, `npm install global-view` and run a basic example, but I get the following error:

> ERROR in ./node_modules/paralleljs/lib/Worker.js
Module not found: Error: Can't resolve 'child_process' in '/home/gabeme/Documents/tinker/test/node_modules/paralleljs/lib'
 @ ./node_modules/paralleljs/lib/Worker.js 1:10-34
 @ ./node_modules/paralleljs/lib/parallel.js
 @ ./node_modules/global-view/es/dataset.js
 @ ./node_modules/global-view/es/index.js
 @ ./src/index.js

GabeMedrash commented on webpack.config.js:
```diff
@@ -0,0 +1,31 @@
+    libraryTarget: 'umd',
+    path: path.resolve(__dirname, 'dist'),
+  },
+  resolve: {
```
Why also do we need this?

hyounesy replied:
not anymore. removed.

hyounesy committed:
eslint: spaced-comment

hyounesy committed:
eslint: indent

hyounesy committed:
eslint: one-var-declaration-per-line, brace-style

hyounesy committed:
eslint: arrow-body-style arrow-parens array-bracket-spacing comma-spacing curly

hyounesy committed:
eslint: arrow-parens: "as-needed"

hyounesy committed:
eslint eqeqeq

hyounesy committed:
eslint: block-scoped-var

hyounesy committed:
eslint: no-mixed-operators

hyounesy committed:
eslint: no-var

hyounesy committed:
eslint: prefer-const

hyounesy committed:
eslint: dot-notation

hyounesy committed:
eslint: no-plusplus

hyounesy committed:
eslint: semi  semi-spacing semi-style

hyounesy committed:
eslint:  no-throw-literal no-useless-escape no-useless-return radix wrap-iife no-multiple-empty-lines padded-blocks quote-props space-before-blocks space-unary-ops prefer-rest-params

hyounesy committed:
eslint: quote-props

hyounesy committed:
eslint: space-infix-ops, prefer-template, template-curly-spacing

hyounesy committed:
eslint: comma-dangle spaced-comment prefer-spread

hyounesy committed:
eslint: no-unneeded-ternary consistent-return import/newline-after-import import/no-extraneous-dependencies ell-last

hyounesy committed:
eslint: function-paren-newline no-extra-bind no-lonely-if  no-new import/no-mutable-exports
parser: eslint-babel

hyounesy committed:
eslint: keyword-spacing key-spacing no-multi-spaces no-extra-semi no-empty one-var

hyounesy committed:
eslint: no-underscore-dangle no-bitwise object-curly-spacing object-curly-newline object-property-newline

hyounesy committed:
eslint no-param-reassign

hyounesy committed:
eslint: object-shorthand no-unused-expressions no-return-assign no-cond-assign

hyounesy committed:
eslint: no-else-return no-alert no-console

hyounesy committed:
fixed benchmark dialog and moved under examples
moved css under examples

hyounesy committed:
eslint: no-nested-ternary no-constant-condition no-fallthrough
had to use case by case eslint-disable for the ones that needed extensive and possibly risky refactoring.

hyounesy committed:
eslint: no-use-before-define max-len

hyounesy committed:
eslint: camelcase, also changed func-names from warn to 0.

hyounesy committed:
eslint: no-multi-assign no-confusing-arrow no-case-declarations

hyounesy committed:
eslint: default-case  guard-for-in  no-extend-native  no-loop-func  no-restricted-properties

hyounesy committed:
eslint: no-unreachable no-new no-shadow no-undef no-unused-vars

hyounesy committed:
eslint: prefer-arrow-callback

hyounesy committed:
eslint: import/prefer-default-export no-prototype-builtins no-restricted-globals no-restricted-syntax prefer-destructuring
disabled the prefer-destructuring in the .eslintrc.js. It is a debated rule, and imho makes the code less readable.

hyounesy committed:
cleaning up: removing intermediate files

hyounesy committed:
added jsdoc to npm packages.

hyounesy committed:
cleaning up examples

hyounesy committed:
reorganized the repo
- moved all example files under src/examples.
- moved all plot test related files under src/test/plot.test

hyounesy committed:
cleaning up: one letter parameter names. function documentation. removing unused functions and comments.

hyounesy committed:
fixed dataset download script: create directory if does not exist

hyounesy committed:
refactoring: changing Dataset from function to class declaration

hyounesy committed:
refactoring: changing from function to class declaration: algorithm, colormap, coordinateSystem, histogramViewer

hyounesy committed:
refactoring: changing from function to class declaration: densityViewer

hyounesy committed:
refactoring: changing function to class declaration: globalview.js
removing dead code and making the long functions modular

hyounesy committed:
cleaning up the globalview.js. Moved Transform class out. moved the event handler functions out of the constructor.

hyounesy committed:
cleaning up: globalview: moved out OPTIONS to separate file and simplified the constructor

hyounesy committed:
cleaning up the remaining modules. converting function constructors to es6 classes.

hyounesy committed:
cleaning up: moving global functions to static class functions

hyounesy committed:
cleaning up package.json:
babel: ignore test and examples
moved dev dependencies

Commit:
remove example code left over from template

Commit:
ignore artifacts

Commit:
accomplish uglification/minification through webpack

Commit:
fix eslint glob pattern

Commit:
remove build artifact from git

hyounesy committed:
run jsdoc
